src/components/selection/common.rs

```diff
@@ -0,0 +1,400 @@
+        }
+    }
+}
+
+impl<P: Problem> Component<P> for CloneSingle {
+    fn execute(&self, problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        selection(self, problem, state)
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct FullyRandom {
+    pub num_selected: u32,
+}
+
+impl FullyRandom {
+    pub fn from_params(num_selected: u32) -> Self {
+        Self { num_selected }
+    }
+
+    pub fn new<P: Problem>(num_selected: u32) -> Box<dyn Component<P>> {
+        Box::new(Self::from_params(num_selected))
+    }
+}
+
+impl<P: Problem> Selection<P> for FullyRandom {
+    fn select<'a>(
+        &self,
+        population: &'a [Individual<P>],
+        rng: &mut Random,
+    ) -> ExecResult<Vec<&'a Individual<P>>> {
+        let mut selection = Vec::new();
+        for _ in 0..self.num_selected {
+            selection.push(population.choose(rng).unwrap());
+        }
+        Ok(selection)
+    }
+}
+
+impl<P: Problem> Component<P> for FullyRandom {
+    fn execute(&self, problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        selection(self, problem, state)
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct RandomWithoutRepetition {
+    pub num_selected: u32,
+}
+
+impl RandomWithoutRepetition {
+    pub fn from_params(num_selected: u32) -> Self {
+        Self { num_selected }
+    }
+
+    pub fn new<P: Problem>(num_selected: u32) -> Box<dyn Component<P>> {
+        Box::new(Self::from_params(num_selected))
+    }
+}
+
+impl<P: Problem> Selection<P> for RandomWithoutRepetition {
+    fn select<'a>(
+        &self,
+        population: &'a [Individual<P>],
+        rng: &mut Random,
+    ) -> ExecResult<Vec<&'a Individual<P>>> {
+        let selection = population
+            .choose_multiple(rng, self.num_selected as usize)
+            .collect();
+        Ok(selection)
+    }
+}
+
+impl<P: Problem> Component<P> for RandomWithoutRepetition {
+    fn execute(&self, problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        selection(self, problem, state)
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct RouletteWheel {
+    pub num_selected: u32,
+    pub offset: f64,
+}
+
+impl RouletteWheel {
+    pub fn from_params(num_selected: u32, offset: f64) -> Self {
+        Self {
+            num_selected,
+            offset,
+        }
+    }
+
+    pub fn new<P: SingleObjectiveProblem>(num_selected: u32, offset: f64) -> Box<dyn Component<P>> {
+        Box::new(Self::from_params(num_selected, offset))
+    }
+}
+
+impl<P: SingleObjectiveProblem> Selection<P> for RouletteWheel {
+    fn select<'a>(
+        &self,
+        population: &'a [Individual<P>],
+        rng: &mut Random,
+    ) -> ExecResult<Vec<&'a Individual<P>>> {
+        let weights = f::proportional_weights(population, self.offset)
+            .wrap_err("population contains invalid objective values")
+            .note("roulette wheel note does work with infinite or negative objective values")?;
```

Comment:
There are possible ways to make this work for negative objective values. I thought I had implemented one of those at some point. Ultimately, we should make this work, either by adapting this implementation or by offering a second roulette wheel variant (though that would be rather unnecessary).

Reply:
I originally removed the handling of negative fitness values because I wasn't sure we were handling all corner cases correctly, and didn't have the time to research this in more detail.

I now have, so please take a look and verify the new implementation.

Reply:
Looks good, thanks!